### PR TITLE
show cody context files as an action in the transcript

### PR DESCRIPTION
**Pull request by sqs**
Instead of showing these in a longer message after Cody's response, the files that Cody read are now shown in a small chip before the response: `7 files read`. Expanding it shows the first step (searching) and each file. This UI can be extended to more kinds of plugins/actions in the future.

image
image

## Test plan

Open Cody. Run a query that requires it to read files. Confirm they show up here.

## App preview:

- [Web](https://sg-web-sqs-cody-contextfiles-as-action.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | -0.00% (-0.46 kb) | -0.00% (-0.46 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/sqs/cody-contextfiles-as-action/compare-report.html;tab=live_object) for a full comparison between the commits e9b550d and 607ecd1 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/cody-ui/src/chat/actions/TranscriptAction.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+export const TranscriptAction: React.FunctionComponent<{
+    title: string | { verb: string; object: string }
+    steps: TranscriptActionStep[]
+    className?: string
+}> = ({ title, steps, className }) => {
+    const [open, setOpen] = useState(false)
+
+    return (
+        <div className={classNames(className, styles.container, open && styles.containerOpen)}>
```
Comment (dominiccooney) [inline client/cody-ui/src/chat/actions/TranscriptAction.tsx:28]: Gave similar feedback to @abeatrix , but HTML has `<details><summary>` for this use case and you get the open/closed toggle with decent accessibility for free. This is fine; let's keep it in mind for next time.

File: client/cody-ui/src/chat/ContextFiles.tsx [hunk 2]
```diff
@@ -15,60 +14,17 @@ export const ContextFiles: React.FunctionComponent<{
...
-    <svg role="img" height={24} width={24} viewBox="0 0 24 24" fill="currentColor">
-        <path d={svgPath} />
-    </svg>
+}> = ({ contextFiles, fileLinkComponent: FileLink, className }) => (
+    <TranscriptAction
+        title={{ verb: 'Read', object: `${contextFiles.length} ${pluralize('file', contextFiles.length)}` }}
+        steps={[
+            { verb: 'Searched', object: 'entire codebase for relevant files', icon: mdiMagnify },
+            ...contextFiles.map(file => ({
+                verb: 'Read',
```
Comment (dominiccooney) [inline client/cody-ui/src/chat/ContextFiles.tsx:23]: I can't tell if the repetition of "Read... Read... Read..." is reassuring or tiring.

**Review by dominiccooney — APPROVED** [review]
Looks great.

You mentioned actions, I am guessing we will want actions have more prominently backgrounds/look like buttons to signal they're interactive. However for supporting context this looks splendid.

File: client/cody-ui/src/chat/ContextFiles.tsx (hunk 2 above)
Comment (sqs) [reply client/cody-ui/src/chat/ContextFiles.tsx:23]: yeah. I was inspired by this, but I think @danielmarquespt @almeidapaulooliveira can do a lot better.
image

File: client/cody-ui/src/chat/actions/TranscriptAction.tsx (hunk 1 above)
Comment (sqs) [reply client/cody-ui/src/chat/actions/TranscriptAction.tsx:28]: ah, yeah, good point. I think this impl will change soon anyway, but that is a very good tip.

**Comment by sqs** [comment]
@dominiccooney: Thanks!

> You mentioned actions, I am guessing we will want actions have more prominently backgrounds/look like buttons to signal they're interactive. However for supporting context this looks splendid.

Actions is probably the wrong term for this, then. I was thinking of this as "a representation of the actions that the assistant took on your behalf" rather than "actions you can perform".

Anyway, easy to change in the code in the future.

File: client/cody-ui/src/chat/ContextFiles.tsx (hunk 2 above)
Comment (almeidapaulooliveira) [reply client/cody-ui/src/chat/ContextFiles.tsx:23]: What about:
'I searched the entire codebase for relevant files and read those:'

And remove the 'Read' from every line?

**Pull request merged**